External Metrics contd

john-delivuk opened the pull request:
- Attempted to isolate and prune some duplicated logic. 
- Cleaned up comments from PR #79 

I was hesitant to start extracting some of the duplicated components. Let me know if you would like this refined further or we can continue to clean up in future work. 

Thanks to @dangoodman5425 and @antoinne85 for everything they've contributed.

pvdvreede commented:
Just noticed the binary was committed: https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/146/files#diff-2fad0bac2a901c821b32c417b2ee77b8 was this by accident?

DirectXMan12 commented:
cc @brancz

brancz commented:
Generally I think having full external metrics support would be nice, but I'd like us to first tackle the existing known scalability issues, that from our recent discussions would significantly change how the adapter works, including this functionality. What do you think @DirectXMan12 ?

antoinne85 commented:
@brancz Thanks for your feedback!

I've been away from this PR for a while now and, admittedly, have no idea which scalability issues you're referring to (I'm not currently following any of the topics in Slack or other areas very closely). So please take the following questions with that in mind.

* Are you referring to scalability issues within the boilerplate or with the way external metrics are implemented in K8S?
* If you're referring to issues within K8S, specifically, are there any rough estimates floating around about when they might get addressed? 

I'm not attempting to apply any pressure at all, just curious about approximate timelines. My general rule-of-thumb on these kinds of things is to not "let perfect be the enemy of good." For example, if the scalability fixes are 2 months away, the time-to-external-metrics-adapter-rework is 2 months plus some additional lead time. If it's six months away, you get the picture.

In the meantime, it _seems_ like there's this external metrics API out there that not many people can actually use because they'd be required to roll their own adapter to Prometheus (or whatever metrics backend they're using).

Now, I've not been involved with the K8S community enough to warrant my having an opinion on the matter, but it seems like (if the code is up-to-quality) it might be worthwhile to go ahead and enable external metrics scenarios. (If for no other reason than the code has already been written, and there's work to be done _after_ the scalability issues are fixed _anyway_ to bring this up-to-speed.)

Again, no negative tone intended. Just trying to get an understanding of things so I can make the proper moves on our end.

john-delivuk commented:
Also preventing an unessecary fork, as well as long as configuration doesn't change, or change much. We can perform these changes in the background.

brancz commented:
Working on all of this is next on our list. The biggest problem I would have with merging this is that the configuration aspect is likely to change dramatically with the planned changes. If if everyone is ok with that, and we list it in the release notes, I'm ok with moving forward with this. What do you think @DirectXMan12 @s-urbaniak ?

s-urbaniak commented:
First, thanks a lot @john-delivuk for reviving these PRs!

I am :+1: on the comments from @brancz. Having said that we could make it clear in the readme that not only big (probably backwards incompatible?) configuration changes are to be expected not only for external (this contribution) but also for the existing custom metrics support.

Regarding the further refactoring we'd have to take external metrics into account anyways. Not having existing code would tempt to "we'll do it later". That way we have some natural pressure to implement external metrics for the future factorings as well.

I'd suggest to do another review sweep throughout the code before merging but it seems to be a good and long awaited addition.

Obviously I am leaving the final word to @DirectXMan12 :-)

DirectXMan12 commented:
I'm fine with the approach of "merge now, disclaimer that it will change later" (don't forget to follow semver ;-) )

DirectXMan12 commented:
Code-changes-wise, I believe I've reviewed a previous version of this, and I trust the other reviewers here (@brancz @s-urbaniak @metalmatze), so I'm fine moving forward with this (don't forget to squash commits into logical chunks, though).

brancz commented:
Sounds and looks good from my side. In that case @john-delivuk could you please do as @DirectXMan12 said and squash these changes into logical chunks, then this is good to go for me :slightly_smiling_face: .

john-delivuk commented:
F*** me. I just rushed the rebase, I'll clean this up tonight when I get home and have time.

john-delivuk commented:
@brancz Let me know if you need anything else!

s-urbaniak commented on pkg/external-provider/errors.go:
```diff
@@ -0,0 +1,27 @@
+// NewOperatorNotSupportedByPrometheusError creates an error that represents the fact that we were requested to service a query that
+// Prometheus would be unable to support.
+func NewOperatorNotSupportedByPrometheusError() error {
+	return errors.New("operator not supported by prometheus")
```
nit: these should be vars in the style of:
```
var (
  ErrOperatorNotSupportedByPrometheus = errors.New("operator not supported by prometheus")
...
)
```

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,109 @@
+
+// overridableSeriesRegistry is a basic SeriesRegistry
+type externalSeriesRegistry struct {
+	mu sync.RWMutex
```
could you document what fields are protected by this?

john-delivuk replied:
Preferred way to document this? Inline comment ?

s-urbaniak replied:
yes, a small comment is totall sufficient :+1:

john-delivuk replied:
@antoinne85 or @DirectXMan12 
From what I can tell we lock the metrics series when we read to prevent read / write issues. Can you just confirm so I don't mislead :)

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func (r *externalSeriesRegistry) filterAndStoreMetrics(result MetricUpdateResult) {
+
```
nit: obsolete newline

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,109 @@
+	converters := result.converters
+
+	if len(newSeriesSlices) != len(converters) {
+		glog.Errorf("need one set of series per converter")
```
This will panic a few lines below if `len(newSeriesSlices) < len(converters)`. We should return with an error from this function instead.

s-urbaniak commented on pkg/external-provider/metric_converter.go:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	for _, val := range toConvert {
+
```
nit: obsolete newline

s-urbaniak commented on pkg/external-provider/metric_namer.go:
```diff
@@ -0,0 +1,159 @@
+// concurrently.  Returned group-resources are "normalized" as per the
+// MetricInfo#Normalized method.  Group-resources passed as arguments must
+// themselves be normalized.
+type MetricNamer interface {
```
This duplicates most of the code from `custom-provider/metric_namer.go`. Is it possible to merge those two implementations? I diff'ed locally and there seems to be some differences, albeit not many.

john-delivuk replied:
So logically I feel like the metric namer should move to `naming`. Opinions ?

s-urbaniak replied:
`naming` sounds like a good choice :+1:

john-delivuk commented:
Thanks for the quick review. Will get to this today I hope. This PR has been open so long it's fallen behind. x.x

s-urbaniak commented:
@john-delivuk thank you so much for the effort! :+1:

s-urbaniak commented:
@john-delivuk another thing, it seems you are vendoring `testify`, as Travis fails with:
```

# github.com/directxman12/k8s-prometheus-adapter/pkg/external-provider
pkg/external-provider/periodic_metric_lister_test.go:8:2: cannot find package "github.com/stretchr/testify/require" in any of:
...
```

Do you mind to vendor it in a separate commit?

john-delivuk commented:
@s-urbaniak One more rebase needed to clean up. If you'd like to look over.

john-delivuk commented:
@brancz not sure if you're interested in nudging this :)

s-urbaniak commented:
@john-delivuk sorry for the delay, please give me one day to give it a final pass.

s-urbaniak commented on pkg/external-provider/errors.go:
```diff
@@ -0,0 +1,21 @@
+var (
+	// ErrNewOperatorNotSupportedByPrometheus creates an error that represents the fact that we were requested to service a query that
+	// Prometheus would be unable to support.
+	ErrNewOperatorNotSupportedByPrometheus = errors.New("operator not supported by prometheus")
```
nit: let's just call this `ErrNotSupportedByPrometheus`

s-urbaniak replied:
(along with the error message `not supported by prometheus`)

s-urbaniak replied:
just a suggestion: doesn't `ErrUnsupportedQuery` make more sense here?

john-delivuk replied:
I think `ErrUnsupportedOperator` going by the error text. Opinion?

anderson4u2 replied:
the more explicit the better init

s-urbaniak commented on pkg/external-provider/errors.go:
```diff
@@ -0,0 +1,21 @@
+
+	// ErrNewOperatorRequiresValues creates an error that represents the fact that we were requested to service a query
+	// that was malformed in its operator/value combination.
+	ErrNewOperatorRequiresValues = errors.New("operator requires values")
```
nit: doesn't `ErrMalformedQuery` make more sense here?

john-delivuk replied:
Your phrasing like the one above points people at the query. Should we have more universal query error and consolidate? I'm happy either way.

s-urbaniak commented on pkg/external-provider/errors.go:
```diff
@@ -0,0 +1,21 @@
+
+	// ErrNewLabelNotSpecified creates an error that represents the fact that we were requested to service a query
+	// that was malformed in its label specification.
+	ErrNewLabelNotSpecified = errors.New("label not specified")
```
nit: let's just call this `ErrLabelNotSpecified`

s-urbaniak commented on pkg/external-provider/errors.go:
```diff
@@ -0,0 +1,21 @@
+
+	// ErrNewOperatorDoesNotSupportValues creates an error that represents the fact that we were requested to service a query
+	// that was malformed in its operator/value combination.
+	ErrNewOperatorDoesNotSupportValues = errors.New("operator does not support values")
```
from the comment there seems to be no difference to `ErrNewOperatorRequiresValues`. Can we be more precise in the godoc?

naming nit: `ErrMalformedQueryUnsupportedValues`

s-urbaniak commented on pkg/external-provider/provider.go:
```diff
@@ -0,0 +1,80 @@
+	"github.com/directxman12/k8s-prometheus-adapter/pkg/naming"
+)
+
+// TODO: Make sure everything has the proper licensing disclosure at the top.
```
please add the LICENSE disclosure ;-)

john-delivuk commented:
@s-urbaniak Changes are updated. Let me know if there is anything else.

s-urbaniak commented:
@john-delivuk awesome, thanks! this is looking great, just one last thing, please add the following header too all introduced files:
```
/*
Copyright 2019 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

john-delivuk commented:
Sorry about that, will do now. I looked at the one file in `pkg/customprovider` that was missing it lol.

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,110 @@
+
+	mapper apimeta.RESTMapper
+
+	metricLister MetricListerWithNotification
```
@john-delivuk I might have missed something, but is this used somewhere?

john-delivuk replied:
I need to dig in, but I believe this is tied to how updates to the Metrics name cache are handled.

s-urbaniak replied:
As commented in https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/146#discussion_r259388559, the constructor of the series registry just installs a notification receiver callback to be invoked when metrics are being listed periodically.

There is no need to have it here as a struct member.

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,110 @@
+		rawMetrics:   map[string]SeriesConverter{},
+	}
+
+	lister.AddNotificationReceiver(registry.filterAndStoreMetrics)
```
I believe we just need it here to receive the callback, but not in the struct.

s-urbaniak commented on pkg/external-provider/external_series_registry.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+// NewExternalSeriesRegistry creates an ExternalSeriesRegistry driven by the data from the provided MetricLister.
+func NewExternalSeriesRegistry(lister MetricListerWithNotification, mapper apimeta.RESTMapper) ExternalSeriesRegistry {
```
same here, I removed `mapper` (also in the callers) and everything compiled happily. Do we need it here?

s-urbaniak commented on pkg/external-provider/series_converter.go:
```diff
@@ -0,0 +1,289 @@
+
+	selectors := []queryPart{}
+	for i := 0; i < len(requirements); i++ {
+		fmt.Println("This is post Requirements", requirements[i].Key(), requirements[i].Values().List(), requirements[i].Operator())
```
do you mind to remove the `fmt.Println`?

s-urbaniak commented on pkg/external-provider/series_converter.go:
```diff
@@ -0,0 +1,289 @@
+	queryBuilder      QueryBuilder
+	seriesFilterer    SeriesFilterer
+	metricNamer       naming.MetricNamer
+	mapper            apimeta.RESTMapper
```
this `mapper` is unsued, we can remove it.

s-urbaniak commented on pkg/external-provider/series_converter.go:
```diff
@@ -0,0 +1,289 @@
+
+	return &seriesConverter{
+		seriesQuery:       prom.Selector(rule.SeriesQuery),
+		mapper:            mapper,
```
see above, `mapper` is not used in `seriesConverter`

Santiclause commented on pkg/external-provider/series_converter.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+func (c *seriesConverter) buildNamespaceQueryPartForExternalSeries(namespace string) (queryPart, error) {
+	namespaceLbl, _ := c.metricNamer.LabelForResource(naming.NsGroupResource)
```
`c.metricNamer.LabelForResource` will fail here due to nil pointer dereferencing, it should be `c.resourceConverter.LabelForResource`

john-delivuk commented:
Quick updated, I've made the requested changes. I'm going to run through some quick testing locally to make sure everything is functioning before I push.

s-urbaniak commented:
@john-delivuk awesome, thank you! I will also test it out once you push the necessary changes. Once done, this will LGTM :+1:

john-delivuk commented:
Just want to confirm everything is kosher before I rebase this. You give me the green light and I'll take care of it.

john-delivuk commented:
tagging @s-urbaniak

s-urbaniak commented:
@john-delivuk This looks fine from my side :+1:

However, I have a small issue, but it might be a misconfiguration too. I took a stab and deployed external metrics, registered prometheus adapter as an api service for external metrics and deployed the following configuration:

```
    externalRules:
    - seriesQuery: 'http_requests_total{namespace!="",pod!=""}'
      resources:
        overrides:
          namespace: {resource: "namespace"}
          pod: {resource: "pod"}
          service: {resource: "service"}
      name:
        matches: "^(.*)_total"
        as: "${1}_per_second"      
      metricsQuery: 'sum(rate(<<.Series>>{<<.LabelMatchers>>}[2m])) by (<<.GroupBy>>)'
```

The query that is executed based on the above config, is the following which seems off:
```
$ kubectl get --raw '/apis/external.metrics.k8s.io/v1beta1/namespaces/default/http_requests_per_second' | jq .
{
  "kind": "ExternalMetricValueList",
  "apiVersion": "external.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/external.metrics.k8s.io/v1beta1/namespaces/default/http_requests_per_second"
  },
  "items": []
}
```
The actual executed prom query is this:
```
I0318 13:21:46.919244       1 api.go:74] GET http://prometheus-operated.default.svc:9090/api/v1/query?query=sum%28rate%28http_requests_per_second%7B%7D%5B2m%5D%29%29+by+%28%29&time=1552915306.917 200 OK
```
url-decoded:
```
sum(rate(http_requests_per_second{}[2m])) by ()
```

john-delivuk commented:
TESTING THE FUNCTIONALITY! How could you!?

I'll dig in today (hopefully), Thanks for this.

Multiply commented:
We're very interested in seeing this merged and released, so let me know if we can do anything to help. (Like testing)

john-delivuk commented:
If anyone is interested in playing with this, I've been pushing builds to my personal dockerhub.

https://hub.docker.com/r/jdelivuk/k8s-prometheus-adapter-amd64

Multiply commented:
I actually have it working. I have yet to create autoscaling to use the rules. Although, I am not using the grouping functionality mentioned above.

Multiply commented:
The only thing I am having a hard time to get working is having external metrics for a specific namespace. My data does not include kubernetes-related information, so I need to add the namespace in, somehow.
I saw https://github.com/DirectXMan12/k8s-prometheus-adapter/issues/167, which seems to request such a feature.

Is there something the code in this PR can do to help in this regard, or will it only be available in the default namespace?

Multiply commented:
I ended up using `metric_relabel_configs` to add the correct namespace to prometheus for now.

I can now get metrics out using the raw API like this:
```
$ kubectl get --raw '/apis/external.metrics.k8s.io/v1beta1/namespaces/production/queue_messages?labelSelector=queue=default,vhost=root'
{"kind":"ExternalMetricValueList","apiVersion":"external.metrics.k8s.io/v1beta1","metadata":{"selfLink":"/apis/external.metrics.k8s.io/v1beta1/namespaces/production/queue_messages"},"items":[{"metricName":"","metricLabels":{},"timestamp":"2019-03-20T14:51:25Z","value":"800"}]}
```

But it doesn't seem like HorizontalPodAutoscaler can access it:
```
E0320 14:49:08.536433       1 horizontal.go:189] failed to compute desired number of replicas based on listed metrics for Deployment/production/queue-default: failed to get queue_messages external metric: unable to get external metric production/queue_messages/&LabelSelector{MatchLabels:map[string]string{queue: default,vhost: root,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: queue_messages.external.metrics.k8s.io is forbidden: User "system:serviceaccount:kube-system:horizontal-pod-autoscaler" cannot list queue_messages.external.metrics.k8s.io in the namespace "production"
I0320 14:49:08.536486       1 event.go:218] Event(v1.ObjectReference{Kind:"HorizontalPodAutoscaler", Namespace:"production", Name:"queue-default", UID:"081f4772-4b1a-11e9-bd8f-02a321a654d2", APIVersion:"autoscaling/v2beta1", ResourceVersion:"32927459", FieldPath:""}): type: 'Warning' reason: 'FailedGetExternalMetric' unable to get external metric production/queue_messages/&LabelSelector{MatchLabels:map[string]string{queue: default,vhost: root,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: queue_messages.external.metrics.k8s.io is forbidden: User "system:serviceaccount:kube-system:horizontal-pod-autoscaler" cannot list queue_messages.external.metrics.k8s.io in the namespace "production"
I0320 14:49:08.536512       1 event.go:218] Event(v1.ObjectReference{Kind:"HorizontalPodAutoscaler", Namespace:"production", Name:"queue-default", UID:"081f4772-4b1a-11e9-bd8f-02a321a654d2", APIVersion:"autoscaling/v2beta1", ResourceVersion:"32927459", FieldPath:""}): type: 'Warning' reason: 'FailedComputeMetricsReplicas' failed to get queue_messages external metric: unable to get external metric production/queue_messages/&LabelSelector{MatchLabels:map[string]string{queue: default,vhost: root,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: queue_messages.external.metrics.k8s.io is forbidden: User "system:serviceaccount:kube-system:horizontal-pod-autoscaler" cannot list queue_messages.external.metrics.k8s.io in the namespace "production"
```

Am I missing something obvious?

john-delivuk commented:
Just an update. I found the issue and patched it.... while along the way hating how half of naming was being handled by `pkg/naming`, and the other half was in `pkg/external-metrics` so pretty decent rewrite coming up. I do want to cycle through some tests tomorrow as I think there are a few edge cases I don't think are being handled correctly

john-delivuk commented:
Quick design question. For `.DiscoveryRule.resource.overrides` this is not really needed for external metrics. I see a few options,
1. Split the DiscoveryRule into 2 different types
2. Document it's unused, have `pkg/naming` just ignore that if specified. 
3. Have it follow suit. 

i'm preferential to 2. Let me know if I'm also missing something, or you would like a different approach.

john-delivuk commented:
@Multiply I'll look into your mentioned issue after these pending changes. I've refactored a lot of this, and will test soon to see if this is still an issue.

john-delivuk commented:
Ok so this change looks big and scary but here is the gist. 
```
1. Remove the functionality duplicated by `pkg/naming`
2. I moved in the query builder to the appropriate sounding place for external queries
3. Removed the namespace section of the query builder. After looking at the `provider.ExternalMetricInfo` it is currently ignorant to the namespace, so I feel like our queries should be as well. If someone would like to specify a namespace they should use a label selector which would translate. As a result, `.resources.overrides` are ignored for any external metric.

I threw up an image on my docker hub mentioned, feel free to try it out. This build feels like it could be the one :)

s-urbaniak commented:
this is looking great @john-delivuk :+1: thanks so much for putting the effort into this! I will this iteration out today, if goes well this is a big LGTM from my side :-)

Multiply commented:
@john-delivuk What is the correct raw path to query for external metrics?

john-delivuk commented:
```kubectl get --raw '/apis/external.metrics.k8s.io/v1beta1/namespaces/[hpaNamespace]/[metricName]?labelSelector=[label]=[targetValue]'```

To view available metrics

```kubectl get --raw '/apis/external.metrics.k8s.io/v1beta1```

s-urbaniak commented:
@john-delivuk PR #179 has just been merged. To make it easier for you, I created the following commit which you can cherry-pick: https://github.com/s-urbaniak/k8s-prometheus-adapter/commit/bbac2bcbd15c147a086d5c95f458c830829bdfcb

john-delivuk commented:
Completed @s-urbaniak

john-delivuk commented:
Is it rebase time?

s-urbaniak commented:
@john-delivuk testing now on my cluster, once successful, we're definitely ready to merge :+1:

s-urbaniak commented:
@john-delivuk awesome work, i just did a sanity test and am getting external metrics :tada: 
```
{
  "kind": "ExternalMetricValueList",
  "apiVersion": "external.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/external.metrics.k8s.io/v1beta1/namespaces/default/http_requests_per_second"
  },
  "items": [
    {
      "metricName": "",
      "metricLabels": {},
      "timestamp": "2019-04-03T14:01:33Z",
      "value": "100099m"
    }
  ]
}
```

s-urbaniak commented:
just for completeness, I also verified both custom metrics and resource metrics:
```
$ kubectl get --raw '/apis/custom.metrics.k8s.io/v1beta1/namespaces/default/pods/example-app-6ff9ff9d-vltfs/http_requests_per_second' | jq .
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/namespaces/default/pods/example-app-6ff9ff9d-vltfs/http_requests_per_second"
  },
  "items": [
    {
      "describedObject": {
        "kind": "Pod",
        "namespace": "default",
        "name": "example-app-6ff9ff9d-vltfs",
        "apiVersion": "/v1"
      },
      "metricName": "http_requests_per_second",
      "timestamp": "2019-04-03T14:17:38Z",
      "value": "100001m"
    }
  ]
}

$ kubectl top pod --all-namespaces 
NAMESPACE     NAME                                   CPU(cores)   MEMORY(bytes)   
default       example-app-6ff9ff9d-vltfs             43m          8Mi             
default       prometheus-adapter-5bc69ff95c-mcm9h    1m           11Mi            
default       vegeta-84fbb5595f-wxrhh                305m         14Mi            
kube-system   coredns-86c58d9df4-shrvd               75m          8Mi             
kube-system   coredns-86c58d9df4-w7bvd               74m          9Mi             
kube-system   etcd-minikube                          24m          48Mi            
kube-system   kube-addon-manager-minikube            0m           4Mi             
kube-system   kube-apiserver-minikube                28m          404Mi           
kube-system   kube-controller-manager-minikube       41m          50Mi            
kube-system   kube-proxy-zp78b                       2m           11Mi            
kube-system   kube-scheduler-minikube                9m           12Mi            
kube-system   storage-provisioner                    0m           13Mi            
monitoring    alertmanager-main-0                    3m           18Mi            
monitoring    alertmanager-main-1                    3m           17Mi            
monitoring    alertmanager-main-2                    3m           17Mi            
monitoring    grafana-5f5677f45d-vgm9m               2m           27Mi            
monitoring    kube-state-metrics-656c58fc5b-ql7dq    1m           31Mi            
monitoring    node-exporter-fwrp7                    3m           19Mi            
monitoring    prometheus-adapter-646885b85b-5hcb5    1m           12Mi            
monitoring    prometheus-k8s-0                       19m          123Mi           
monitoring    prometheus-k8s-1                       20m          129Mi           
monitoring    prometheus-operator-7cb68545c6-g87hv   9m           22Mi            

$ kubectl top node 
NAME       CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
minikube   931m         23%    1506Mi          20%       
```

zacharysierakowski commented:
@s-urbaniak I am trying to use the prometheus adapter helm charts (https://github.com/helm/charts/tree/master/stable/prometheus-adapter) but I'm assuming there are adjustments that need to be made for v0.5.0 since the current version set in the values.yaml file is v0.4.1. Do you know if anyone is working on updating these for v0.5.0 or if there are particular things I could do manually to make them work?

Sorry if this is the wrong place to ask.

s-urbaniak commented:
@zacharysierakowski I am not aware of any work done for helm. All changes in 0.5.0 should be backwards compatible with 0.4.1. What 0.5.0 introduces, is a new configuration section `externalRules`, so effectively the helm chart should still work.

zacharysierakowski commented:
Thanks for such a quick reply! Are there docs on the `externalRules` configuration section I can refer to? When I switched 0.4.1 to 0.5.0 and deployed it, everything seemed green and good to go. But when running `kubectl get --raw '/apis/external.metrics.k8s.io/v1beta1'`, I didn't get anything. I'm assuming that it's b/c I missed that new configuration section.

s-urbaniak commented:
@zacharysierakowski external metrics (as well as custom metrics) are something you have to configure yourself. If the config is missing, that api won't be enabled. I made a configuration sample here https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/146#issuecomment-473915775.

As mentioned above, I'll create additional documentation. For the sake of this already long PR, please open a new issue if you have additional questions :-)

zacharysierakowski commented:
@s-urbaniak Okay great. Thank you very much for you help!
